File: 0904-fruit-into-baskets/0904-fruit-into-baskets.py
class Solution:
    def totalFruit(self, fruits: List[int]) -> int:
        total = 0
        count = 0
        fruit_types = Counter()
        left = right = 0
        while right < len(fruits):
            if fruit_types[fruits[right]] == 0:
                count += 1
            fruit_types[fruits[right]] += 1
            
            while count > 2:
                fruit_types[fruits[left]] -= 1
                if fruit_types[fruits[left]] == 0:
                    count -= 1
                left += 1
            
            total = max(total, right - left + 1)
            right +=1
        return total
        

File: 0904-fruit-into-baskets/0904-fruit-into-baskets.js
/**
 * @param {number[]} fruits
 * @return {number}
 */
var totalFruit = function(fruits) {
    let max = 0, size = 0, left = 0, map = {};
    fruits.forEach((fruit, i) => {
        map[fruit] ? map[fruit]++ : (map[fruit] = 1, size++)
        while (size > 2) {
            const tail = fruits[left]
            if (--map[tail] === 0) delete map[tail], size--
            left++
        }
        max = Math.max(max, i - left + 1)
    })
    return max;
};